Google.Api.Gax.Grpc/ClientBuilderBase.cs

```diff
@@ -0,0 +1,266 @@
+            }
+            return new DefaultCallInvoker(channel);
+        }
+
+        /// <summary>
+        /// Obtains channel credentials synchronously. Override this method in a concrete builder type if more
+        /// credential mechanisms are supported.
+        /// </summary>
+        protected virtual ChannelCredentials GetChannelCredentials()
+        {
+            if (ChannelCredentials != null)
+            {
+                return ChannelCredentials;
+            }
+            if (TokenAccessMethod != null)
+            {
+                return new DelegatedTokenAccess(TokenAccessMethod).ToChannelCredentials();
+            }
+            GoogleCredential unscoped =
+                CredentialsPath != null ? GoogleCredential.FromFile(CredentialsPath) :
+                JsonCredentials != null ? GoogleCredential.FromJson(JsonCredentials) :
+                GoogleCredential.GetApplicationDefault();
+            return unscoped.CreateScoped(Scopes ?? GetDefaultScopes()).ToChannelCredentials();
+        }
+
+        /// <summary>
+        /// Obtains channel credentials asynchronously. Override this method in a concrete builder type if more
+        /// credential mechanisms are supported.
+        /// </summary>
+        protected async virtual Task<ChannelCredentials> GetChannelCredentialsAsync(CancellationToken cancellationToken)
+        {
+            if (ChannelCredentials != null)
+            {
+                return ChannelCredentials;
+            }
+            if (TokenAccessMethod != null)
+            {
+                return new DelegatedTokenAccess(TokenAccessMethod).ToChannelCredentials();
+            }
+            GoogleCredential unscoped =
+                CredentialsPath != null ? GoogleCredential.FromFile(CredentialsPath) : // TODO: Use an async method when one is available
+                JsonCredentials != null ? GoogleCredential.FromJson(JsonCredentials) :
+                await GoogleCredential.GetApplicationDefaultAsync().ConfigureAwait(false);
+            return unscoped.CreateScoped(Scopes ?? GetDefaultScopes()).ToChannelCredentials();
+        }
+
+        /// <summary>
+        /// Returns the default scopes for this builder type, used if no scopes are otherwise specified.
+        /// </summary>
+        protected abstract IReadOnlyList<string> GetDefaultScopes();
+
+        /// <summary>
+        /// Returns the endpoint for this builder type, used if no endpoint is otherwise specified.
+        /// </summary>
+        protected abstract ServiceEndpoint GetDefaultEndpoint();
+
+        /// <summary>
+        /// Returns the channel pool to use when no other options are specified. This method is not called unless
+        /// <see cref="CanUseChannelPool"/> returns true, so if a channel pool is unavailable, override that property
+        /// to return false and throw an exception from this method.
+        /// </summary>
+        protected abstract ChannelPool GetChannelPool();
```

Comment:
Alternatively remove `CanUseChannelPool` and make this `GetChannelPool()` method return `null` if a channel-pool can't be used?

Reply:
Hmm. Maybe. Although that makes derived class implementations a bit weirder, in terms of either having to repeat the existing logic or have a second member anyway. I think in reality all clients are likely to have a channel pool they *could* use when appropriate, so most of the logic would be about suitability.

Reply:
OK, fine to leave as it is :)